src/debug.cc

```diff
@@ -91,6 +372,215 @@ ResyncDebugLog(FILE *newFile)
     TheLog.file_ = newFile;
 }
 
+/* DebugChannel */
+
+void
+DebugChannel::stopEarlyMessageCollection()
+{
+    if (const auto toLog = releaseEarlyMessages())
+        logSaved(*toLog);
+}
+
+void
+Debug::ForgetSaved()
+{
+    auto &module = Module();
+    (void)module.cacheLogChannel.releaseEarlyMessages();
+    (void)module.stderrChannel.releaseEarlyMessages();
+    (void)module.syslogChannel.releaseEarlyMessages();
+}
+
+void
+Debug::SwanSong()
+{
+    Module().swanSong();
+}
+
+void
+DebugChannel::logSaved(const DebugMessages &messages)
+{
+    const auto writtenEarlier = written;
+    for (const auto &message: messages) {
+        if (Debug::Enabled(message.header.section, message.header.level))
+            log(message.header, message.body);
+    }
+    const auto writtenNow = written - writtenEarlier;
+    debugs(0, 5, "wrote " << writtenNow << " out of " <<
+           messages.size() << " early messages to " << name);
+}
+
+bool
+DebugChannel::saveMessage(const DebugMessageHeader &header, const std::string &body)
+{
+    if (!earlyMessages)
+        return false; // we have stopped saving early messages
+
+    if (header.level > EarlyMessagesLevel)
+        return false; // this message is not important enough to save
+
+    // There should not be a lot of messages because EarlyMessagesLevel is
+    // small, but we limit their accumulation just in case.
+    const DebugMessages::size_type limit = 1000;
+    DebugMessages::size_type purged = 0;
+    if (earlyMessages->size() >= limit) {
+        DebugMessages doomedMessages;
+        earlyMessages->swap(doomedMessages);
+        purged = doomedMessages.size(); // before handleOverflow() may change it
+        handleOverflow(doomedMessages);
+    }
+
+    earlyMessages->emplace_back(header, body);
+
+    if (purged) {
+        // This debugs() should come _after_ we save the early message above,
+        // preserving the original event and LogMessage() order, like log().
+        debugs(0, DBG_CRITICAL, "ERROR: Too many early important messages. " <<
+               "Purged " << purged << " from " << name);
+    }
+
+    return true;
+}
+
+void
+DebugChannel::writeToStream(FILE &destination, const DebugMessageHeader &header, const std::string &body)
+{
+    fprintf(&destination, "role=%s # %s%s| %s\n",
+            header.role_XXX,
+            debugLogTime(header.timestamp),
+            debugLogKid(),
+            body.c_str());
+    noteWritten(header);
+}
+
+void
+DebugChannel::noteWritten(const DebugMessageHeader &header)
+{
+    ++written;
+    lastWrittenRecordNumber = header.recordNumber;
```

Comment:
I think that lastWrittenRecordNumber should not decrease - we ought to add a condition here. I guess this may happen for nested messages, for example, after  DebugChannel::saveMessage() generates the critical "Too many early important messages".

Reply:
Are you implying that, in the current PR code, lastWrittenRecordNumber may decrease (because of nested messages) despite Channel::log() checks that are meant to prevent such decreases? If the answer is yes, does that imply that a channel may log records out of order as well?